src/main/java/com/mongodb/spark/sql/connector/config/MongoConfig.java

```diff
@@ -17,191 +17,198 @@
    */
-  public static MongoConfig createOutputConfig(final CaseInsensitiveStringMap options) {
-    return new MongoConfig(options, UsageMode.OUTPUT);
-  }
+  String CONNECTION_STRING_DEFAULT = "mongodb://localhost:27017/";
 
-  private final CaseInsensitiveStringMap originals;
-  private final CaseInsensitiveStringMap values;
-  private final UsageMode usageMode;
-  private transient MongoClientFactory mongoClientFactory;
-
-  private MongoConfig(final CaseInsensitiveStringMap originals, final UsageMode usageMode) {
-    this.originals = originals;
-    this.usageMode = usageMode;
-
-    String ignorePrefix = usageMode == UsageMode.INPUT ? MONGO_OUTPUT_PREFIX : MONGO_INPUT_PREFIX;
-    String overridePrefix = usageMode == UsageMode.INPUT ? MONGO_INPUT_PREFIX : MONGO_OUTPUT_PREFIX;
-
-    Map<String, String> usageSpecificSettings = new HashMap<>();
-    List<String> defaults = new ArrayList<>();
-    List<String> overrides = new ArrayList<>();
-
-    originals.keySet().stream()
-        .filter(k -> k.startsWith(MONGO_PREFIX))
-        .forEach(
-            k -> {
-              if (k.startsWith(overridePrefix)) {
-                overrides.add(k);
-              } else if (!k.startsWith(ignorePrefix)) {
-                defaults.add(k);
-              }
-            });
-
-    defaults.forEach(
-        k -> usageSpecificSettings.put(k.substring(MONGO_PREFIX.length()), originals.get(k)));
-    overrides.forEach(
-        k -> usageSpecificSettings.put(k.substring(overridePrefix.length()), originals.get(k)));
-
-    this.values = new CaseInsensitiveStringMap(usageSpecificSettings);
-  }
+  /**
+   * The database name config
+   *
+   * <p>{@value}
+   */
+  String DATABASE_NAME_CONFIG = "database";
 
   /**
-   * Convert the configuration for use when inputting data
+   * The collection name config
    *
-   * @return the MongoConfig to use.
+   * <p>{@value}
    */
-  public MongoConfig toInputMongoConfig() {
-    if (this.usageMode == UsageMode.INPUT) {
-      return this;
-    }
-    return new MongoConfig(originals, UsageMode.INPUT);
-  }
+  String COLLECTION_NAME_CONFIG = "collection";
+
+  /** @return the options for this MongoConfig instance */
+  Map<String, String> getOptions();
 
   /**
-   * Convert the configuration for use when outputting data
+   * Create a new {@link MongoConfig} instance with the extra options.
+   *
+   * <p>Existing configurations may be overwritten by the new options.
    *
-   * @return the MongoConfig to use.
+   * @param options the context specific options.
+   * @return a new MongoConfig
    */
-  public MongoConfig toOutputMongoConfig() {
-    if (this.usageMode == UsageMode.OUTPUT) {
-      return this;
-    }
-    return new MongoConfig(originals, UsageMode.OUTPUT);
+  MongoConfig withOptions(Map<String, String> options);
+
+  /** @return the original options for this MongoConfig instance */
+  Map<String, String> getOriginals();
+
+  /** @return the read config */
+  ReadConfig toReadConfig();
```

Comment:
You can't mutate the config.  Users can passing override options eg:

```
        df.write()
            .format("mongodb")
            .mode("Overwrite")
            .option("myOption", "myValue");
```

But that will produce a new configuration. All original values are left the same - so that when you call toReadConfig() or toWriteConfig() then specific usage values will override the default ones.  Eg.  `"read.myOption"`in the read context will override any default `myOption` configurations.

This allows users to set a single configuration for the allows default values and usage specific overrides.